7645-hud-util-2026

ttoomey opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

## Description
* Adds HudUtility2026 with minor adjustments from 2024
* Adds spec for new utility class
* Does **not** update new code to reference the new class
* Does **not** include proposed new "current" API

### Questions
Let's discuss how to handle retired elements

## Type of change
 New feature

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [x] I have performed a self-review of my code
- [ ] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] I have added spec tests (or not applicable)

ttoomey committed:
7645-hud-util-2026

ttoomey commented:
Note for review, the changes are minor, see diff from previous year

```diff
--- lib/util/concerns/hud_lists_2024.rb 2025-05-10 23:47:29.823930003 +0000
+++ lib/util/concerns/hud_lists_2026.rb 2025-05-10 23:46:41.953317009 +0000
@@ -7,7 +7,7 @@
 # frozen_string_literal: true
 
 # THIS FILE IS GENERATED, DO NOT EDIT DIRECTLY
-module Concerns::HudLists2024
+module Concerns::HudLists2026
   extend ActiveSupport::Concern
   class_methods do
     # 1.1
@@ -710,8 +710,6 @@
         44 => 'HUD: CoC - Joint Component TH/RRH',
         45 => 'VA: Grant Per Diem - Case Management/Housing Retention',
         46 => 'Local or Other Funding Source',
-        47 => 'HUD: ESG - CV',
-        48 => 'HUD: HOPWA - CV',
         49 => 'HUD: CoC - Joint Component RRH/PSH [Deprecated]',
         50 => 'HUD: HOME',
         51 => 'HUD: HOME (ARP)',
@@ -719,6 +717,7 @@
         53 => 'HUD: ESG - RUSH',
         54 => 'HUD: Unsheltered Special NOFO',
         55 => 'HUD: Rural Special NOFO',
+        56 => 'HUD: CoC Builds',
       }.freeze
     end
 
@@ -1187,6 +1186,21 @@
       _translate(referral_results, id, reverse, raise_on_missing: raise_on_missing)
     end
 
+    # 4.21
+    def sexes
+      {
+        0 => 'Female',
+        1 => 'Male',
+        8 => "Client doesn't know",
+        9 => 'Client prefers not to answer',
+        99 => 'Data not collected',
+      }.freeze
+    end
+
+    def sex(id, reverse = false, raise_on_missing: false)
+      _translate(sexes, id, reverse, raise_on_missing: raise_on_missing)
+    end
+
     # 5.03.1
     def data_collection_stages
       {
@@ -2009,6 +2023,7 @@
         7 => 'Shallow Subsidy',
         8 => 'Returning Home',
         9 => 'Rapid Resolution',
+        10 => 'Healthcare Navigation',
       }.freeze
     end
 
@@ -2016,7 +2031,7 @@
       _translate(ssvf_services_options, id, reverse, raise_on_missing: raise_on_missing)
     end
 
-    # W2.2
+    # V2.3
     def hopwa_financial_assistance_options
       {
         1 => 'Rental assistance',
@@ -2580,10 +2595,10 @@
         'AmIndAKNative' => 'American Indian, Alaska Native, or Indigenous',
         'Asian' => 'Asian or Asian American',
         'BlackAfAmerican' => 'Black, African American, or African',
-        'HispanicLatinaeo' => 'Hispanic/Latina/e/o',
-        'MidEastNAfrican' => 'Middle Eastern or North African',
         'NativeHIPacific' => 'Native Hawaiian or Pacific Islander',
         'White' => 'White',
+        'HispanicLatinaeo' => 'Hispanic/Latina/e/o',
+        'MidEastNAfrican' => 'Middle Eastern or North African',
         'RaceNone' => "Doesn't know, prefers not to answer, or not collected",
       }.freeze
     end
```

```diff
--- lib/data/2024_hud_lists.json        2024-11-22 22:00:05.758334010 +0000
+++ lib/data/2026_hud_lists.json        2025-05-10 16:26:10.477453003 +0000
@@ -600,14 +600,6 @@
         "description": "Local or Other Funding Source"
       },
       {
-        "key": 47,
-        "description": "HUD: ESG - CV"
-      },
-      {
-        "key": 48,
-        "description": "HUD: HOPWA - CV"
-      },
-      {
         "key": 49,
         "description": "HUD: CoC - Joint Component RRH/PSH [Deprecated]"
       },
@@ -634,6 +626,10 @@
       {
         "key": 55,
         "description": "HUD: Rural Special NOFO"
+      },
+      {
+        "key": 56,
+        "description": "HUD: CoC Builds"
       }
     ]
   },
@@ -3344,6 +3340,32 @@
     ]
   },
   {
+    "name": "Sex",
+    "code": "4.21",
+    "values": [
+      {
+        "key": 0,
+        "description": "Female"
+      },
+      {
+        "key": 1,
+        "description": "Male"
+      },
+      {
+        "key": 8,
+        "description": "Client doesn't know"
+      },
+      {
+        "key": 9,
+        "description": "Client prefers not to answer"
+      },
+      {
+        "key": 99,
+        "description": "Data not collected"
+      }
+    ]
+  },
+  {
     "name": "EarlyExitReason",
     "code": "R17.A",
     "values": [
@@ -4274,6 +4296,10 @@
       {
         "key": 9,
         "description": "Rapid Resolution"
+      },
+      {
+        "key": 10,
+        "description": "Healthcare Navigation"
       }
     ]
   },
```

eanders commented on lib/data/2026_hud_lists.json:
For posterity, this is the diff from 2024
```diff
{
         "key": 46,
         "description": "Local or Other Funding Source"
-      },
-      {
-        "key": 47,
-        "description": "HUD: ESG - CV"
-      },
-      {
-        "key": 48,
-        "description": "HUD: HOPWA - CV"
       },
       {
         "key": 49,
@@ -634,6 +626,10 @@
       {
         "key": 55,
         "description": "HUD: Rural Special NOFO"
+      },
+      {
+        "key": 56,
+        "description": "HUD: CoC Builds"
       }
     ]
   },
@@ -3340,6 +3336,32 @@
       {
         "key": 3,
         "description": "Unsuccessful referral: provider rejected"
+      }
+    ]
+  },
+  {
+    "name": "Sex",
+    "code": "4.21",
+    "values": [
+      {
+        "key": 0,
+        "description": "Female"
+      },
+      {
+        "key": 1,
+        "description": "Male"
+      },
+      {
+        "key": 8,
+        "description": "Client doesn't know"
+      },
+      {
+        "key": 9,
+        "description": "Client prefers not to answer"
+      },
+      {
+        "key": 99,
+        "description": "Data not collected"
       }
     ]
   },
@@ -4274,6 +4296,10 @@
       {
         "key": 9,
         "description": "Rapid Resolution"
+      },
+      {
+        "key": 10,
+        "description": "Healthcare Navigation"
       }
     ]
   },
```

gigxz replied:
@ttoomey @eanders  Last time we added the suffix ` [Deprecated]` to labels  for funding sources that were deprecated (12 and 49). I think we should continue that practice. HUD updated the guidance on this to be clearer this time around:
> Retired funding sources must remain on relevant projects created prior to FY 2026 but should not be available to be added to new projects.


So it would probably be best to have 2 lists, `funding_sources` (all funding sources including retired, used for labeling funder numbers), and `funding_sources_current` (currently available funding sources, used for picklist on new funder creation).


I think we can probably tackle the second part of that on https://github.com/open-path/Green-River/issues/7049, so my suggested change here would be to add back the retired funders and indicate them as deprecated.

gigxz replied:
We're also missing a new list for new data Element V10 ("MentalHeathConsultation")
<img width="568" alt="Screenshot 2025-05-12 at 9 45 59 AM" src="https://github.com/user-attachments/assets/1e9cf824-4d6b-468d-a699-52a5440f2752" />

ttoomey replied:
thanks @gigxz 
- I'll preserve the retired elements and add the missing v10, sorry I missed that

gigxz replied:
no prob thank you!

eanders commented on lib/util/concerns/hud_lists_2026.rb:
Diff from 2024
```diff
# THIS FILE IS GENERATED, DO NOT EDIT DIRECTLY
-module Concerns::HudLists2024
+module Concerns::HudLists2026
   extend ActiveSupport::Concern
   class_methods do
     # 1.1
@@ -710,8 +710,6 @@
         44 => 'HUD: CoC - Joint Component TH/RRH',
         45 => 'VA: Grant Per Diem - Case Management/Housing Retention',
         46 => 'Local or Other Funding Source',
-        47 => 'HUD: ESG - CV',
-        48 => 'HUD: HOPWA - CV',
         49 => 'HUD: CoC - Joint Component RRH/PSH [Deprecated]',
         50 => 'HUD: HOME',
         51 => 'HUD: HOME (ARP)',
@@ -719,6 +717,7 @@
         53 => 'HUD: ESG - RUSH',
         54 => 'HUD: Unsheltered Special NOFO',
         55 => 'HUD: Rural Special NOFO',
+        56 => 'HUD: CoC Builds',
       }.freeze
     end
 
@@ -1187,6 +1186,21 @@
       _translate(referral_results, id, reverse, raise_on_missing: raise_on_missing)
     end
 
+    # 4.21
+    def sexes
+      {
+        0 => 'Female',
+        1 => 'Male',
+        8 => "Client doesn't know",
+        9 => 'Client prefers not to answer',
+        99 => 'Data not collected',
+      }.freeze
+    end
+
+    def sex(id, reverse = false, raise_on_missing: false)
+      _translate(sexes, id, reverse, raise_on_missing: raise_on_missing)
+    end
+
     # 5.03.1
     def data_collection_stages
       {
@@ -2009,6 +2023,7 @@
         7 => 'Shallow Subsidy',
         8 => 'Returning Home',
         9 => 'Rapid Resolution',
+        10 => 'Healthcare Navigation',
       }.freeze
     end
 
@@ -2016,7 +2031,7 @@
       _translate(ssvf_services_options, id, reverse, raise_on_missing: raise_on_missing)
     end
 
-    # W2.2
+    # V2.3
     def hopwa_financial_assistance_options
       {
         1 => 'Rental assistance',
@@ -2580,10 +2595,10 @@
         'AmIndAKNative' => 'American Indian, Alaska Native, or Indigenous',
         'Asian' => 'Asian or Asian American',
         'BlackAfAmerican' => 'Black, African American, or African',
-        'HispanicLatinaeo' => 'Hispanic/Latina/e/o',
-        'MidEastNAfrican' => 'Middle Eastern or North African',
         'NativeHIPacific' => 'Native Hawaiian or Pacific Islander',
         'White' => 'White',
+        'HispanicLatinaeo' => 'Hispanic/Latina/e/o',
+        'MidEastNAfrican' => 'Middle Eastern or North African',
         'RaceNone' => "Doesn't know, prefers not to answer, or not collected",
       }.freeze
     end
```

eanders commented on lib/util/hud_utility_2026.rb:
Diff from 2024
```diff
# frozen_string_literal: true
 
-module HudUtility2024
+module HudUtility2026
   include ::Concerns::HudValidationUtil
-  include ::Concerns::HudLists2024
+  include ::Concerns::HudLists2026
 
   module_function
 
@@ -200,7 +200,7 @@
 
   def residential_project_type_numbers_by_codes(*codes)
     codes = codes.flatten # Take either array, or multiple parameters
-    codes.map { |code| HudUtility2024.residential_project_type_numbers_by_code[code] }.
+    codes.map { |code| HudUtility2026.residential_project_type_numbers_by_code[code] }.
       flatten.
       uniq.
       sort.
@@ -679,7 +679,7 @@
       'HUD: Unsheltered Special NOFO' => [54],
       'HUD: Rural Special NOFO' => [55],
       'HUD: HUD-VASH' => [20],
-      'HUD: PFS' => [HudUtility2024.funding_source('HUD: Pay for Success', true, raise_on_missing: true)], # Pay for Success
+      'HUD: PFS' => [HudUtility2026.funding_source('HUD: Pay for Success', true, raise_on_missing: true)], # Pay for Success
       'HUD: HOME' => [50, 51],
     }
   end
@@ -721,43 +721,43 @@
   def service_types_provided_map
     {
       141 => {
-        list: ::HudUtility2024.path_services_options,
+        list: ::HudUtility2026.path_services_options,
         label_method: :path_services,
       },
       142 => {
-        list: ::HudUtility2024.rhy_services_options,
+        list: ::HudUtility2026.rhy_services_options,
         label_method: :rhy_services,
       },
       143 => {
-        list: ::HudUtility2024.hopwa_services_options,
+        list: ::HudUtility2026.hopwa_services_options,
         label_method: :hopwa_services,
       },
       144 => {
-        list: ::HudUtility2024.ssvf_services_options,
+        list: ::HudUtility2026.ssvf_services_options,
         label_method: :ssvf_services,
       },
       151 => {
-        list: ::HudUtility2024.hopwa_financial_assistance_options,
+        list: ::HudUtility2026.hopwa_financial_assistance_options,
         label_method: :hopwa_financial_assistance,
       },
       152 => {
-        list: ::HudUtility2024.ssvf_financial_assistance_options,
+        list: ::HudUtility2026.ssvf_financial_assistance_options,
         label_method: :ssvf_financial_assistance,
       },
       161 => {
-        list: ::HudUtility2024.path_referral_options,
+        list: ::HudUtility2026.path_referral_options,
         label_method: :path_referral,
       },
       200 => {
-        list: ::HudUtility2024.bed_night_options,
+        list: ::HudUtility2026.bed_night_options,
         label_method: :bed_night,
       },
       210 => {
-        list: ::HudUtility2024.voucher_tracking_options,
+        list: ::HudUtility2026.voucher_tracking_options,
         label_method: :voucher_tracking,
       },
       300 => {
-        list: ::HudUtility2024.moving_on_assistance_options,
+        list: ::HudUtility2026.moving_on_assistance_options,
         label_method: :moving_on_assistance,
       },
     }.freeze
@@ -773,15 +773,15 @@
   def service_sub_types_provided_map
     {
       3 => {
-        list: ::HudUtility2024.ssvf_sub_type3s,
+        list: ::HudUtility2026.ssvf_sub_type3s,
         label_method: :ssvf_sub_type3,
       },
       4 => {
-        list: ::HudUtility2024.ssvf_sub_type4s,
+        list: ::HudUtility2026.ssvf_sub_type4s,
         label_method: :ssvf_sub_type4,
       },
       5 => {
-        list: ::HudUtility2024.ssvf_sub_type5s,
+        list: ::HudUtility2026.ssvf_sub_type5s,
         label_method: :ssvf_sub_type5,
       },
     }
```

eanders commented on spec/lib/util/hud_utility_2026_spec.rb:
```diff
@@ -0,0 +1,354 @@
+  #
+  # Testing strategy:
+  # 1. Load the JSON file containing the HUD data lists/codes
+  # 2. For each lookup method in, test both the forward lookup (code to description)
```
```suggestion
  # 2. For each lookup method in the HUD lists file, test both the forward lookup (code to description)
```

eanders reviewed (approved):
Thanks @ttoomey, I think this covers it!

ttoomey committed:
Merge branch 'release-163' into 7645-hud-util-2026

ttoomey committed:
add v10, preserve retired fields

ttoomey committed:
comment

ttoomey committed:
ugh

ttoomey committed:
7645-hud-util-2026

ttoomey commented:
@gigxz I made adjustments per our convo today, does this seem correct to you?
- removed all deprecated / retired elements from the HUD 2026 json/hud list
- created a new HudUtility2026Deprecations module that defines both gender; it also redefines fields that have deprecated options (funding_source_current and ssvf_financial_assistance)
- adjusted HudUtility2026 so that funding_sources_current and ssvf_financial_assistance_options_current are the actual 2026 definitions
- ssvf_financial_assistance_options include deprecated /retired values
- also fixed incorrect hopwa code in 2024 json/list

gigxz commented:
@ttoomey sounds good to me. Since you moved gender to the deprecations concern, we should also move the R3 sexual_orientations list there, which is also removed in 2026. Sorry I missed that last time. LGTM after that!

gigxz commented:
It looks like there is a test failure too

ttoomey committed:
alphabetize order or race/ethnicity

ttoomey committed:
move option to deprecated mixin
